### ISSUE TITLE
JSDoc and typescript declaration file (*.d.ts)

### ISSUE THREAD

**Issue by WoonchanCho** [dcmjs-org/dcmjs#163]
Hi @pieper and other maintainers,
Thank you so much for initiating and maintaining this great project.
I am trying to test this library out as much as I can in our team's projects and enjoy using it now.

This is not a production issue at all but a kind of improvement suggestion.
As you know, the source code itself is not fully documented.
It would be great to have JsDoc comments at the JavaScript source level.

This will help a lot to users who use a modern editor that provides an IntelliSense with the JsDocs comments.
Even more than that, it will also make creating & maintaining the typescript declaration files (*.d.ts) extremely easy.

I will be happy to add the JsDoc comment to all the active JavaScript sources in the project.
For sure, this will be (should be) no harm to the functionalities of the library because this work will only add comments package.json for the two DevDependencies: JsDoc and TypeScript (Typescript will be used only for creating declaration files)

My only concern is that I'm not a native English speaker, there might be some awkward expression used.
If I add those JsDoc comments, can you review them and update if there is anything weird?

Thank you so much.

**Comment by pieper** [comment]
Hi @WoonchanCho and @hackermd - these sound like great improvements to me.  The core code is pretty stable at this point so now would be a good time to put in some more documentation and ts help.  I'd be happy to review and edit 👍

**Comment by hackermd** [comment]
Great! @WoonchanCho I have type annoations for the `sr` subpackage. We could then start from there. I will prepare a pull request.

**Comment by WoonchanCho** [comment]
Sounds great! Then I will wait till @hackermd's pull request is merged into the master branch (or a new branch for this work) and begin working after it.

**Comment by WoonchanCho** [comment]
Hi @hackermd, I just wanted to check the status on the type annotation for the sr subpackage. If you are busy for something or need to take more time I might begin to work and you can add on your work on it.
I'm opened to anything so feel free to let me know.

**Comment by hackermd** [comment]
@WoonchanCho I have a type declarations file. It's not yet complete and I have yet tested its distribution via the npm. I will push what I have and then we can discuss outstanding TODO items. OK?

**Comment by kresli** [comment]
any updates on this? TS support would be very much appreciated.

**Comment by pieper** [comment]
No, I don't think anyone has had a chance to work on it and @hackermd has moved to a new position.  Anyone who wants to pick up the effort is welcome.

**Comment by kresli** [comment]
we could generate `.d.ts` easily by running [dts-gen](https://github.com/microsoft/DefinitelyTyped-tools/tree/main/packages/dts-gen) but that would only work if typescripts can infer types automatically. To help typescript we could write JSDOC. That would be much easier then keep .d.ts manually updated. What is your preference? JSDOC + autogenerated .d.ts  OR hardcoded .d.ts? Here is an example of autogenerated .d.ts. in current state

<img width="884" alt="image" src="https://github.com/user-attachments/assets/cf1c34ec-6221-4884-af26-1dce0cb58ead">

**Comment by pieper** [comment]
I don't really have any experience with ts so I'll defer to others, but anything that simplifies/automates the process would be good.  If it did double-duty to help generate API documentation that sounds like a win.

**Comment by lk101** [comment]
> we could generate `.d.ts` easily by running [dts-gen](https://github.com/microsoft/DefinitelyTyped-tools/tree/main/packages/dts-gen) but that would only work if typescripts can infer types automatically. To help typescript we could write `JSDOC`. That would be much easier then keep .`d.ts` manually updated. What is your preference? `JSDOC` + autogenerated `.d.ts` OR hardcoded `.d.ts`? Here is an example of autogenerated `.d.ts`. in current state
> 
> <img alt="image" width="884" src="https://private-user-images.githubusercontent.com/3826097/360674420-cf1c34ec-6221-4884-af26-1dce0cb58ead.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU3MzUwODMsIm5iZiI6MTc0NTczNDc4MywicGF0aCI6Ii8zODI2MDk3LzM2MDY3NDQyMC1jZjFjMzRlYy02MjIxLTQ4ODQtYWYyNi0xZGNlMGNiNThlYWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MjdUMDYxOTQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2NiNzU5ZTZmMmRjMWY4OTE4YmU5ZTNlM2NiZjUyNjI5YmJhNWMyN2I3NDZmYWYyMTQ0MWM0ZDc0YTZhZTQxMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.OE5AoPHVXIVX279__ZKqoucBu-Z-Giw2Sedvz2pfmRw">

I can assist in writing part of the `d.ts` files. If the author @pieper requires, I could also attempt to migrate the entire project to TypeScript, though the latter would likely require a significant amount of time. However, regarding libraries based on TypeScript files, I don't have hands-on experience with using tools like `tsc` to generate their d.ts and JS files. Therefore, there might be some obstacles in this aspect.

**Comment by lk101** [comment]
Besides that, I also need to learn how to use Yarn (since I’ve been using pnpm all along), but that shouldn’t take too much time.
